nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2BrowseNodeGenerator.java

```diff
@@ -47,34 +46,45 @@
 
   @Override
   public List<BrowsePaths> computeAssetPaths(final Asset asset, @Nullable final Component component) {
+    String remote = component == null ? getRemoteForMetadataAsset(asset.name()) : getRemoteForPackage(asset.name());
+    List<BrowsePaths> paths = Lists.newArrayList(computeComponentPaths(asset, component).iterator());
+    paths.addAll(BrowsePaths
+        .fromPaths(Lists.newArrayList(
+            Splitter.on(DIVIDER).omitEmptyStrings().split(getRelativePath(asset.name(), remote)).iterator()),
+            true));
+    return paths;
+  }
+
+  @Override
+  public List<BrowsePaths> computeComponentPaths(final Asset asset, @Nullable final Component component) {
+    List<String> pathParts = new ArrayList<>();
+    String remote = component == null ? getRemoteForMetadataAsset(asset.name()) : getRemoteForPackage(asset.name());
+    if (!remote.isEmpty()) {
+      pathParts.add(getRemoteWithoutPrefix(remote));
+    }
     if (component != null) {
-      List<BrowsePaths> paths = computeComponentPaths(asset, component);
-      String assetName = getAssetNameWithoutRemotePrefix(asset);
-      assetName = assetName.contains(DIVIDER) ? assetName.substring(assetName.lastIndexOf(DIVIDER) + 1) : assetName;
-      BrowsePaths.appendPath(paths, assetName);
-      return paths;
+      pathParts.addAll(Lists.newArrayList(Splitter.on('.').omitEmptyStrings().split(component.name()).iterator()));
+      pathParts.add(component.version());
     }
+    return BrowsePaths.fromPaths(pathParts, true);
+  }
 
-    return BrowsePaths.fromPaths(Collections.singletonList(getAssetNameWithoutRemotePrefix(asset)), false);
+  private String getRemoteForMetadataAsset(String path) {
+    int endIndex = StringUtils.lastOrdinalIndexOf(path, DIVIDER, 1);
+    return endIndex != -1 ? path.substring(0, endIndex): StringUtils.EMPTY;
```

Comment:
trivial: try to check positive case instead of revert condition

Reply:
Better leave it, because record like endIndex == -1 ? StringUtils.EMPTY: path.substring(0, endIndex); is really unclear